internal/openfga/client.go

```diff
@@ -0,0 +1,138 @@
+package openfga
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	openfga "github.com/openfga/go-sdk"
+	"github.com/openfga/go-sdk/client"
+	"github.com/openfga/go-sdk/credentials"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+)
+
+type Config struct {
+	ApiScheme   string
+	ApiHost     string
+	StoreID     string
+	ApiToken    string
+	AuthModelID string
+	Debug       bool
+}
+
+func NewConfig(apiScheme, apiHost, storeID, apiToken, authModelID string, debug bool) *Config {
+	c := new(Config)
+
+	c.ApiScheme = apiScheme
+	c.ApiHost = apiHost
+	c.StoreID = storeID
+	c.ApiToken = apiToken
+	c.AuthModelID = authModelID
+	c.Debug = debug
+
+	return c
+}
+
+type Client struct {
+	c *client.OpenFgaClient
+}
+
+func (c *Client) APIClient() *client.OpenFgaClient {
+	return c.c
+}
+
+func (c *Client) WriteModel(ctx context.Context, model []byte) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	authModel := new(client.ClientWriteAuthorizationModelRequest)
+
+	if err := json.Unmarshal(model, authModel); err != nil {
+		return "", err
+	}
+
+	data, err := c.c.WriteAuthorizationModelExecute(
+		c.c.WriteAuthorizationModel(ctx).Body(*authModel),
+	)
+
+	if err != nil {
+		return "", err
+	}
+
+	return data.GetAuthorizationModelId(), nil
+}
+
+func (a *Client) ListObjects(ctx context.Context, user string, relation string, objectType string) ([]string, error) {
+	r := a.APIClient().OpenFgaApi.ListObjects(ctx)
+	body := &openfga.ListObjectsRequest{
+		User:     user,
+		Relation: relation,
+		Type:     objectType,
+	}
+	r = r.Body(*body)
+	objectsResponse, _, err := a.APIClient().OpenFgaApi.ListObjectsExecute(r)
+	if err != nil {
+		return nil, err
+	}
+
+	allowedObjs := make([]string, len(objectsResponse.GetObjects()))
+	for i, p := range objectsResponse.GetObjects() {
+		// remove the "{objectType}:" prefix from the response
+		allowedObjs[i] = p[len(fmt.Sprintf("%s:", objectType)):]
+	}
+
+	return allowedObjs, nil
+}
+
+func (a *Client) Check(ctx context.Context, user string, relation string, object string) (bool, error) {
+	r := a.APIClient().OpenFgaApi.Check(ctx)
+	body := openfga.NewCheckRequest(
+		openfga.TupleKey{
+			User:     openfga.PtrString(user),
+			Relation: openfga.PtrString(relation),
+			Object:   openfga.PtrString(object),
+		},
+	)
+	r = r.Body(*body)
+
+	check, _, err := a.APIClient().OpenFgaApi.CheckExecute(r)
+	if err != nil {
+		return false, err
+	}
+
+	return check.GetAllowed(), nil
+}
+
+func NewClient(cfg *Config) *Client {
```

Comment:
given there is some substantial logic in here let's add logger, monitor and tracer